### PR TITLE
chore: rewrite semantics to not have pattern match on structure

### PR DESCRIPTION
**Pull request by bollu**
### Issues:

This works around https://github.com/leanprover/lean4/issues/5388, and removes uses of `simp (config := { ground := true }) ..`. This makes our proof states cleaner, since we no longer see proof states involving `Decidable.rec ...`. Peeled from #168 

### Testing:

What tests have been run? Did `make all` succeed for your changes? Was
conformance testing successful on an Aarch64 machine?

`make all` succeeded, conformance succeeded.

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by shigoel** [comment]
BTW, `AddWithCarry` is also used here:
```
./Insts/DPR/Add_sub_shifted_reg.lean:    let (result, pstate) := AddWithCarry operand1 operand2 carry
./Insts/DPR/Add_sub_carry.lean:    let (result, pstate) := AddWithCarry operand1 operand2 (read_flag PFlag.C s)
```

File: Arm/Insts/DPI/Add_sub_imm.lean [hunk 1]
```diff
@@ -24,19 +24,21 @@ def exec_add_sub_imm (inst : Add_sub_imm_cls) (s : ArmState) : ArmState :=
                         else
                           imm <<< 12
     let operand1      := read_gpr datasize inst.Rn s
-    let (carry_in, operand2)
+    let carryInAndOperand2
```
Comment (shigoel) [inline Arm/Insts/DPI/Add_sub_imm.lean:27]: Could we leave a note (and an issue) to remind us to restore this definition to the original form after https://github.com/leanprover/lean4/issues/5388 is fixed?

File: Arm/Insts/DPI/Add_sub_imm.lean [hunk 2]
```diff
@@ -24,19 +24,21 @@ def exec_add_sub_imm (inst : Add_sub_imm_cls) (s : ArmState) : ArmState :=
...
                       := if sub_op then
                           (1#1, ~~~imm)
                         else
                           (0#1, imm)
+    let carry := carryInAndOperand2.fst
+    let operand2 := carryInAndOperand2.snd
     let operand2         := BitVec.zeroExtend datasize operand2
-    let (result, pstate) := AddWithCarry operand1 operand2 carry_in
+    let resultAndPState := AddWithCarry operand1 operand2 carry
```
Comment (alexkeizer) [inline Arm/Insts/DPI/Add_sub_imm.lean:35]: This pattern-match should be fine, right? IIUC, it was only matching on an if that caused problems

**Comment by bollu** [comment]
> BTW, `AddWithCarry` is also used here:
> 
> ```
> ./Insts/DPR/Add_sub_shifted_reg.lean:    let (result, pstate) := AddWithCarry operand1 operand2 carry
> ./Insts/DPR/Add_sub_carry.lean:    let (result, pstate) := AddWithCarry operand1 operand2 (read_flag PFlag.C s)
> ```


@shigoel these are okay, because their arguments don't come from a `if...then...else`. Consider the diff:

```
-    let (carry_in, operand2)
+    let carryInAndOperand2
                       := if sub_op then
                           (1#1, ~~~imm)
                         else
                           (0#1, imm)
+    let carry := carryInAndOperand2.fst
+    let operand2 := carryInAndOperand2.snd
     let operand2         := BitVec.zeroExtend datasize operand2
-    let (result, pstate) := AddWithCarry operand1 operand2 carry_in
+    let (result, pstate) := AddWithCarry operand1 operand2 carry
```

The issue is that we were pattern matching to extract `(carry_in, operand2)` from the `if ... then ... else` which elaborated poorly. The culprit is this pattern match, not so much the use of `AddWithCarry`. `AddWithCarry` would then try to use the `carry`, but any function that used `carry` would have lead to a problem.

File: Arm/Insts/DPI/Add_sub_imm.lean (hunk 1 above)
Comment (bollu) [reply Arm/Insts/DPI/Add_sub_imm.lean:27]: Done!

File: Arm/Insts/DPI/Add_sub_imm.lean (hunk 2 above)
Comment (bollu) [reply Arm/Insts/DPI/Add_sub_imm.lean]: Fixed, undone.

**Comment by bollu** [comment]
@shigoel updated.

**Review by shigoel — APPROVED** [review]

**Pull request merged**